noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,191 @@
+// Minimal NFT implementation with `AuthWit` support that allows minting in public-only and transfers in both public
+// and private.
+contract NFT {
+    #[aztec(event)]
+    struct Transfer {
+        from: AztecAddress,
+        to: AztecAddress,
+        token_id: Field,
+    }
+
+    #[aztec(storage)]
+    struct Storage {
+        // The symbol of the NFT
+        symbol: SharedImmutable<FieldCompressedString>,
+        // The name of the NFT
+        name: SharedImmutable<FieldCompressedString>,
+        // The admin of the contract
+        admin: PublicMutable<AztecAddress>,
+        // Addresses that can mint
+        minters: Map<AztecAddress, PublicMutable<bool>>,
+        // Contains the NFTs owned by each address in private.
+        private_nfts: Map<AztecAddress, PrivateSet<NFTNote>>
+        // Contains number of NFTs owned by each address in public.
+        public_balances: Map<AztecAddress, PublicMutable<U128>>,
+        // A map from token ID to a boolean indicating if the NFT exists.
+        nft_exists: Map<Field, PublicMutable<bool>>,
+        // A map from token ID to the public owner of the NFT.
+        public_owners: Map<Field, PublicMutable<AztecAddress>>,
+        // Contains the NFTs that are waiting to be shielded.
+        pending_shields: PrivateSet<TransparentNote>,
+    }
+
+    #[aztec(public)]
+    #[aztec(initializer)]
+    fn constructor(admin: AztecAddress, name: str<31>, symbol: str<31>, decimals: u8) {
+        assert(!admin.is_zero(), "invalid admin");
+        storage.admin.write(admin);
+        storage.minters.at(admin).write(true);
+        storage.name.initialize(FieldCompressedString::from_string(name));
+        storage.symbol.initialize(FieldCompressedString::from_string(symbol));
+    }
+
+    #[aztec(public)]
+    fn set_admin(new_admin: AztecAddress) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.admin.write(new_admin);
+    }
+
+    #[aztec(public)]
+    fn set_minter(minter: AztecAddress, approve: bool) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.minters.at(minter).write(approve);
+    }
+
+    #[aztec(public)]
+    fn mint_public(to: AztecAddress, token_id: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
+        // assert the token_id does not exist yet and mint
+        ...
+    }
+
+    #[aztec(public)]
+    fn mint_private(token_id: Field, secret_hash: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
```

Comment:
Not being able to unset minters sounds pretty scary (what if the minter keys leak?) but going with shared mutable makes sense to me.

But this same argument applies to standard token as well (I have copied these shielding/minting functions to the NFT). It feels like we should finalize the discussion of what flows we want to support (will put some 🧠 to it today).


## Off-topic thought
In the standard token another team introduced [privately_mint_private_note](https://github.com/AztecProtocol/aztec-packages/blob/4a840cfab755046e2f713e542680dc1604d5bc9a/noir-projects/noir-contracts/contracts/token_contract/src/main.nr#L223) function because they found it valuable and then they don't use shared mutable to assert minter but instead schedule a call to public. There it makes sense because they need to bump the total supply of the token. I just realized that it will be very common that the token will be bridged from L1 and hence will have total supply tracked in ERC20 on L1. If we decide to have a version of a token without total supply this would allow us to truly privately mint on L2.

This seems very powerful because it could allow you to fully conceal the when a given "shield on L1" is claimed on L2. Currently you are always forced to leak the total supply bump which makes it trivial to determine when a given shield was claimed on L2 (because the amounts are not quantized).

Reply:
I'm fine with minting in public initially, but yeah using shared mutable for the minters seems reasonable. The alternatives are minting in public, which might be sort of fine for erc20 (see below) but seems subpar here, or using private mutable, which has large DA costs. The only issue with Shared Mutable is that it introduces delays when removing minters from the set, which wouldn't help much in your key leakage example, but maybe we can find a workaround to this limitation.

Re. your other thought, total supply is unfortunately a critical requirement in many use cases, including most of DeFi (e.g. math on liquidity pools requires you to reason about how much of the pool a given user owns), so removing it from a fungible token would be quite the endeavor.

Reply:
> Re. your other thought, total supply is unfortunately a critical requirement in many use cases, including most of DeFi (e.g. math on liquidity pools requires you to reason about how much of the pool a given user owns), so removing it from a fungible token would be quite the endeavor.

I feel like this is necessary for public DeFi so don't you think it would be fine if we had something like public_total_supply and simply not track the private one?

For the bridged tokens the total supply would anyway not track the total supply (only the bridged token supply) so I have doubts that revealing the info here makes sense.

I mean maybe the idea is 💩 but I feel like we have not sufficiently considered this because it's for sure that due to private state some of the design patterns of Aztec DeFi will be different from Ethereum's. Might be another thing to try to spec out...

Reply:
>The only issue with Shared Mutable is that it introduces delays when removing minters from the set, which wouldn't help much in your key leakage example, but maybe we can find a workaround to this limitation.

Is the main concern here that the malicious minter could be minting longer due to the delay?

Reply:
Created [an issue](https://github.com/AztecProtocol/aztec-packages/issues/8416) for the total supply thing since having the claim flow be essentially public is bothering me a lot.